scs-user-api/src/auth.rs

```diff
@@ -1,161 +1,161 @@
-        .take(30)
-        .map(char::from)
-        .collect(),
-    }
-  }
-
-  pub fn encode(&self) -> String {
-    base64::encode_config(format!("{}-{}", self.user_id, self.token), base64::URL_SAFE)
-  }
-
-  pub fn decode(value: &str) -> Option<Self> {
-    let bytes = base64::decode_config(value, base64::URL_SAFE).ok()?;
-    let string = String::from_utf8(bytes).ok()?;
-    let (user_id, token) = string
-      .split_once('-')
-      .and_then(|(id, token)| Some((id.parse::<i32>().ok()?, token.to_string())))?;
-    Some(AccessToken { user_id, token })
-  }
-}
-
-fn bearer_auth_value(v: &header::HeaderValue) -> Option<&str> {
-  v.to_str()
-    .ok()
-    .and_then(|v| v.strip_prefix("Bearer ").or_else(|| v.strip_prefix("bearer ")))
-}
-
-impl FromRequest for AccessToken {
-  type Error = crate::error::Error;
-  type Future = Pin<Box<dyn Future<Output = Result<Self, Self::Error>>>>;
-
-  fn from_request(req: &actix_web::HttpRequest, _: &mut actix_http::Payload) -> Self::Future {
-    let auth = req
-      .headers()
-      .get(header::AUTHORIZATION)
-      .and_then(bearer_auth_value)
-      .and_then(AccessToken::decode);
-
-    let db = req.app_data::<web::Data<db::Database>>().unwrap().clone();
-    Box::pin(async move {
-      let auth = auth.with(StatusCode::UNAUTHORIZED)?;
-      if db::tokens::verify(db.get_ref(), auth.user_id(), &auth.token)
-        .await
-        .internal()?
-      {
-        Ok(auth)
-      } else {
-        Err(StatusCode::UNAUTHORIZED.into())
-      }
-    })
-  }
-}
-
-impl serde::Serialize for AccessToken {
-  fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-  where
-    S: serde::Serializer,
-  {
-    <String as serde::Serialize>::serialize(&self.encode(), serializer)
-  }
-}
-
-impl<'de> serde::Deserialize<'de> for AccessToken {
-  fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-  where
-    D: serde::Deserializer<'de>,
-  {
-    use serde::de::Error;
-    let string = <String as serde::Deserialize<'de>>::deserialize(deserializer)?;
-    AccessToken::decode(&string).ok_or_else(|| Error::custom("invalid access token"))
-  }
-}
+use crate::error::FailWith;
+use actix_http::StatusCode;
+use actix_web::http::header;
+use actix_web::{post, web, FromRequest, HttpResponse, Responder, Result};
+use rand::{distributions::Alphanumeric, thread_rng, Rng};
+use std::{future::Future, pin::Pin};
+
+#[derive(Debug, serde::Deserialize)]
+pub struct TokenQuery {
+  pub code: String,
+  pub redirect_uri: String,
+}
+
+#[derive(Debug, serde::Serialize)]
+pub struct TokenResponse {
+  pub token: AccessToken,
+}
+
+#[post("/token")]
+pub async fn create_token(
+  db: web::Data<db::Database>,
+  client: web::Data<reqwest::Client>,
+  secret: web::Data<ClientSecret>,
+  query: web::Query<TokenQuery>,
+) -> Result<impl Responder> {
+  use crate::ex::*;
+
+  // request authorization
+  log::info!("[authorization] {} {} {}", secret.0, query.code, query.redirect_uri);

```

Comment:
🔢

Reply:
Should I take out all of those debug logs?